File: Java/Implementations/Strategy/BipedalMovement.java
public class BipedalMovement implements Movement {
	private MovementSpeed movementSpeed;

	public BipedalMovement(MovementSpeed movementSpeed) {
		this.movementSpeed = movementSpeed;
	}

	public String moveTo(Position from, Position to) {
		double distance = from.distanceTo(to);
		double speed = this.movementSpeed.speed();
		double time = distance / speed;

		return this.movementSpeed.verb() + " from " + from + " to " + to
			+ ",\na distance of " + Utils.roundTo(distance, 2)
			+ ",\nat a speed of " + Utils.roundTo(speed, 2)
			+ ",\nwhich will take " + Utils.roundTo(time, 2) + " seconds";
	}

	public enum MovementSpeed {
		WALK("walking", 1),
		RUN("running", 3);

		private double speed;
		private String verb;

		private MovementSpeed(String verb, double speed) {
			this.verb = verb;
			this.speed = speed;
		}

		private double speed() {
			return this.speed;
		}

		private String verb() {
			return this.verb;
		}
	}
}

File: Haskell/Implementations/Strategy/BipedalMovement.hs
module BipedalMovement
    ( bipedalMovement
    , MovementSpeed(..)
    ) where

import Movement (Movement, distanceBetween)
import Util (roundTo)

data MovementSpeed
    = Walking
    | Running

movementSpeedToDouble :: MovementSpeed -> Double
movementSpeedToDouble m = case m of
    Walking -> 1
    Running -> 3

bipedalMovement :: MovementSpeed -> Movement
bipedalMovement ms from to = mconcat
    [ "walking from "
    , show from
    , " to "
    , show to
    , ",\na distance of "
    , round2 distance
    , ",\nat a speed of "
    , round2 speed
    , ",\nwhich will take "
    , round2 time
    , " seconds"
    ]
  where
    distance = distanceBetween from to
    speed    = movementSpeedToDouble ms
    time     = distance / speed
    round2   = show . roundTo 2
